src/PrestaShopBundle/ApiPlatform/NormalizationMapper.php

```diff
@@ -76,4 +81,91 @@ public function mapNormalizedData(mixed &$normalizedData, array &$context): void
         // Mapping is only done once, so we unset it for next recursive calls
         unset($context[self::NORMALIZATION_MAPPING]);
     }
+
+    /**
+     * Update normalization mapping when it contains indexes, for example:
+     *
+     *     '[categoriesInformation][categoriesInformation][@index][id]' => '[categories][@index][categoryId]',
+     *
+     * is transformed into:
+     *
+     *     '[categoriesInformation][categoriesInformation][0][id]' => '[categories][0][categoryId]',
+     *     '[categoriesInformation][categoriesInformation][1][id]' => '[categories][1][categoryId]',
+     *
+     * depending on the computed array length from the normalized data.
+     */
+    protected function updateMappingIndexes(mixed $normalizedData, array &$normalizationMapping): void
+    {
+        // Copy mapping that will  be modified by reference, to leave the loop one untouched while we do modifications
+        $updatedNormalizationMapping = $normalizationMapping;
+        foreach ($updatedNormalizationMapping as $originPath => $targetPath) {
+            $this->transformMappingPath($normalizedData, $originPath, $targetPath, $normalizationMapping);
+        }
+    }
+
+    /**
+     * Transform one of the mapping path if it contains indexes, the original path containing placeholders is cleaned.
+     */
+    protected function transformMappingPath(mixed $normalizedData, string $originPath, string $targetPath, array &$normalizationMapping): void
+    {
+        $propertyPath = new PropertyPath($originPath);
+        $indexes = [];
+        for ($i = 0; $i < $propertyPath->getLength(); ++$i) {
+            $property = $propertyPath->getElement($i);
+            if (in_array($property, $indexes)) {
+                throw new InvalidArgumentException(sprintf('You cannot use the same index twice %s', $property));
+            }
+
+            if (str_starts_with($property, '@')) {
+                $indexes[] = $property;
+            }
+        }
+
+        // No indexes detected nothing to do
+        if (empty($indexes)) {
```

Comment:
(NIT) Avoid `empty()` method (some SA explode with this because it has weird side effects).

```suggestion
        if ($indexes === []) {
```

Reply:
Hmm, really? Most of the times it does a better job than the `===` IMHO because it also catches null values, and it checks array sub parts (like isset), which is very often something you want to do as well

So, more often than not I rely on `empty` Which side effects are you thinking about?

Reply:
Good articles to read about this : 
- https://dev.to/klnjmm/never-use-empty-function-in-php-4pb0
- https://www.beberlei.de/post/when_to_use_empty_in_php_i_say_never
- https://azizdev.medium.com/why-i-avoid-using-empty-in-php-a-closer-look-12765e87542c
- https://localheinz.com/articles/2023/05/10/avoiding-empty-in-php/

Reply:
Sure, but as long as you are aware of these subtleties to me, it's fine I don't really see the interest to do, for example:

```php
if (isset($array['field']) && is_array($array['field']) && count($array['field']) === 0) {
}
```

when this is almost equivalent

```php
if (!empty($array['field'])) {
}
```

As long as you know which type your data is, usually you're fine. And if the variable can be of multiple unexpected types it probably means the problem was the code before the `empty`

Reply:
The case where I don't use it though is when there are actual difference, for example if I want to check the presence of a field in the array even if it's empty (for partial update let's say) In this case `array_key_exists` is more suited